### PR TITLE
from_u32_unchecked: check validity, and fix UB in Wtf8

### LINKED ISSUE #72760: OsString::from_wide (in Windows OsStringExt) is unsound
The following program causes UB:
```rust
use std::os::windows::ffi::{OsStrExt, OsStringExt};
use std::ffi::{OsStr, OsString};

fn main() {
    let base = "a\té \u{7f}💩\r";
    let mut base: Vec<u16> = OsStr::new(base).encode_wide().collect();
    base.push(0xD800);
    let _res = OsString::from_wide(&base);
}
```
Miri says:
```
error: Undefined Behavior: type validation failed: encountered 0x0000d800, but expected a valid unicode codepoint
   --> /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libcore/char/convert.rs:102:5
    |
102 |     transmute(i)
    |     ^^^^^^^^^^^^ type validation failed: encountered 0x0000d800, but expected a valid unicode codepoint
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
            
    = note: inside `std::char::from_u32_unchecked` at /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libcore/char/convert.rs:102:5
    = note: inside `std::sys_common::wtf8::Wtf8Buf::push_code_point_unchecked` at /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/sys_common/wtf8.rs:204:26
    = note: inside `std::sys_common::wtf8::Wtf8Buf::from_wide` at /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/sys_common/wtf8.rs:194:21
    = note: inside `<std::ffi::OsString as std::os::windows::ffi::OsStringExt>::from_wide` at /home/r/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/sys/windows/ext/ffi.rs:101:44
note: inside `main` at wtf8.rs:8:16
   --> wtf8.rs:8:16
    |
8   |     let _res = OsString::from_wide(&base);
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The problem is this code:

https://github.com/rust-lang/rust/blob/96dd4690c3aa70ec312448c3f2d50e6dc6fb87df/src/libstd/sys_common/wtf8.rs#L293-L305

This calls `push_code_point_unchecked` unless the new code point is in `0xDC00..=0xDFFF`, but what about surrogates in `0xD800..0xDC00`?

This code is unchanged since its introduction in https://github.com/rust-lang/rust/commit/c5369ebc7f4791c4e291951751b8964052c7a523. I am not sure what the intended safety contract of `push_code_point_unchecked` is. That method is not marked `unsafe` but clearly should be -- it calls `char::from_u32_unchecked`. So my guess is the safety precondition is that `CodePoint` must not be part of a surrogate pair, but the thing is, `push` calls it without actually ensuring that condition. The condition it *does* ensure is that the codepoint is not in `0xDC00..=0xDFFF`, but that does not help.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/72760

### CODE REVIEW

File: src/libcore/char/convert.rs [hunk 1]
```diff
@@ -99,7 +99,7 @@ pub fn from_u32(i: u32) -> Option<char> {
 #[inline]
 #[stable(feature = "char_from_unchecked", since = "1.5.0")]
 pub unsafe fn from_u32_unchecked(i: u32) -> char {
-    transmute(i)
+    if cfg!(debug_assertions) { char::from_u32(i).unwrap() } else { transmute(i) }
```
Comment (ollie27) [inline src/libcore/char/convert.rs:102]: `from_u32` ends up calling `from_u32_unchecked` so won't this result in an infinite recursion?

Comment (RalfJung) [reply src/libcore/char/convert.rs:102]: *oops* looks like I did not dig deep enough when I looked at the call graph here.

**Comment by RalfJung** [comment]
@Mark-Simulacrum I fixed the endless recursion.

**Comment by RalfJung** [comment]
r? @Mark-Simulacrum 
(sorry for double-ping)

**Comment by Mark-Simulacrum** [comment]
@bors r+

No worries!

**Comment by RalfJung** [comment]
This actually caused CI to fail in https://github.com/rust-lang/rust/pull/72736#issuecomment-636022427
```
---- sys_common::wtf8::tests::wtf8_as_str stdout ----
thread 'sys_common::wtf8::tests::wtf8_as_str' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/char/convert.rs:102:33

---- sys_common::wtf8::tests::wtf8_code_points stdout ----
thread 'sys_common::wtf8::tests::wtf8_code_points' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/char/convert.rs:102:33

...
```
@bors r-

Our wtf8 code seems to have UB.^^

**Comment by RalfJung** [comment]
Yeah this code seems wrong:

https://github.com/rust-lang/rust/blob/master/src/libstd/sys_common/wtf8.rs#L293-L305

this calls `push_code_point_unchecked` unless the new code point is in `0xDC00..=0xDFFF`, but what about surrogates in `0xD800..0xDC00`?

This code is unchanged since its introduction in https://github.com/rust-lang/rust/commit/c5369ebc7f4791c4e291951751b8964052c7a523. @rust-lang/libs does anyone know what the intended safety contract of `push_code_point_unchecked` is? That method is not marked `unsafe` but clearly should be -- it calls `char::from_u32_unchecked`. So my guess is the safety precondition is that `CodePoint` must not be part of a surrogate pair, but the thing is, `push` calls it without actually ensuring that condition.

**Comment by RalfJung** [comment]
I opened an issue for the UB: https://github.com/rust-lang/rust/issues/72760
This PR is blocked on fixing that first.

**Comment by RalfJung** [comment]
@SimonSapin I implemented the fix you suggested here.

File: src/libcore/char/methods.rs [hunk 2]
```diff
@@ -1673,3 +1635,60 @@ impl char {
...
+        3
+    } else {
+        4
+    }
+}
+
+/// Encodes a raw u32 value as UTF-8 into the provided byte buffer,
+/// and then returns the subslice of the buffer that contains the encoded character.
+///
+/// Unlike `char::encode_utf8`, this method can be called on codepoints in the surrogate range.
```
Comment (Mark-Simulacrum) [inline src/libcore/char/methods.rs:1634]: What does it do on those codepoints -- does it produce valid UTF-8? Based on the UB issue it sounds like yes. Can you add that to this comment? Maybe a brief "note that `char` has a validity invariant that excludes surrogate values" too.

**Comment by Mark-Simulacrum** [comment]
r=me with comment added, unless you want to wait for @SimonSapin's review in particular (but AFAICT this is just moving some code around so I feel comfortable approving)

File: src/libcore/char/methods.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/char/methods.rs:1634]: Is it valid UTF-8 when a surrogate pair gets encoded? I don't know.

Comment (SimonSapin) [reply src/libcore/char/methods.rs:1634]: Good point.

This function codes a code point in [generalized UTF-8](https://simonsapin.github.io/wtf-8/#generalized-utf8). For surrogate code points, the resulting byte sequence is not well-formed UTF-8 and therefore not valid `str`. The return type should be changed to `&mut [u8]` and the `from_utf8_unchecked` call moved to `char::encode_utf8`.

**Comment by RalfJung** [comment]
I added comments but tried to avoid claiming that this is valid UTF-8... what do you think?

File: src/libcore/char/methods.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/char/methods.rs]: > The return type should be changed to &mut [u8] and the from_utf8_unchecked call moved to char::encode_utf8.

Done.

File: src/libcore/char/methods.rs [hunk 3]
```diff
@@ -1673,3 +1615,100 @@ impl char {
...
+/// Panics if the buffer is not large enough.
+/// A buffer of length 2 is large enough to encode any `char`.
+#[unstable(feature = "char_internals", reason = "exposed only for libstd", issue = "none")]
+#[doc(hidden)]
+#[inline]
+pub fn encode_utf16_raw(mut code: u32, dst: &mut [u16]) -> &mut [u16] {
+    // SAFETY: each arm checks whether there are enough bits to write into
+    unsafe {
+        if (code & 0xFFFF) == code && !dst.is_empty() {
+            // The BMP falls through (assuming non-surrogate, as it should)
```
Comment (RalfJung) [inline src/libcore/char/methods.rs:1696]: "assuming non-surrogate" here sounds suspicious?

Comment (SimonSapin) [reply src/libcore/char/methods.rs:1696]: I think this parenthesis should be removed. No surrogate is a valid assumption for `char::encode_utf16` but not here. Encoding a surrogate code point as a single 16-bit code unit is the desired behavior.

Comment (RalfJung) [reply src/libcore/char/methods.rs]: Okay, I removed it.

**Review by Mark-Simulacrum — APPROVED** [review]
r=me

**Pull request merged**